applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -438,88 +455,179 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
                     }
-                    if(aux.size1() != dofs.size() || aux.size2() != mNumberOfRomModes) {
-                        aux.resize(dofs.size(), mNumberOfRomModes,false);
+
+                    // Calculate condition contribution
+                    if (condition_is_active) {
+                        Condition::DofsVectorType dofs;
+                        it->GetDofList(dofs, r_current_process_info);
+                        pScheme->CalculateSystemContributions(*it, LHS_Contribution, RHS_Contribution, EquationId, r_current_process_info);
+                        const auto &geom = it->GetGeometry();
+                        if(PhiElemental.size1() != dofs.size() || PhiElemental.size2() != mNumberOfRomModes) {
+                            PhiElemental.resize(dofs.size(), mNumberOfRomModes,false);
+                        }
+                        if(aux.size1() != dofs.size() || aux.size2() != mNumberOfRomModes) {
+                            aux.resize(dofs.size(), mNumberOfRomModes,false);
+                        }
+                        RomAuxiliaryUtilities::GetPhiElemental(PhiElemental, dofs, geom, mMapPhi);
+                        noalias(aux) = prod(LHS_Contribution, PhiElemental);
+                        double h_rom_weight = it->GetValue(HROM_WEIGHT);
+                        noalias(tempA) += prod(trans(PhiElemental), aux) * h_rom_weight;
+                        noalias(tempb) += prod(trans(PhiElemental), RHS_Contribution) * h_rom_weight;
                     }
-                    RomAuxiliaryUtilities::GetPhiElemental(PhiElemental, dofs, geom, mMapPhi);
-                    noalias(aux) = prod(LHS_Contribution, PhiElemental);
-                    double h_rom_weight = it->GetValue(HROM_WEIGHT);
-                    noalias(tempA) += prod(trans(PhiElemental), aux) * h_rom_weight;
-                    noalias(tempb) += prod(trans(PhiElemental), RHS_Contribution) * h_rom_weight;
                 }
-            }
 
-            #pragma omp critical
-            {
-                noalias(Arom) += tempA;
-                noalias(brom) += tempb;
-            }
+                #pragma omp critical
+                {
+                    noalias(Arom) += tempA;
+                    noalias(brom) += tempb;
+                }
 
+            }
         }
+        
 
         KRATOS_INFO_IF("ROMBuilderAndSolver", (this->GetEchoLevel() > 0 && rModelPart.GetCommunicator().MyPID() == 0)) << "Build time: " << assembling_timer.ElapsedSeconds() << std::endl;
         KRATOS_INFO_IF("ROMBuilderAndSolver", (this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished parallel building" << std::endl;
 
         //solve for the rom unkowns dunk = Arom^-1 * brom
         Vector dxrom(mNumberOfRomModes);
         const auto solving_timer = BuiltinTimer();
-        MathUtils<double>::Solve(Arom, dxrom, brom);
+        if (mSolveWithQR){
+            QR<double, row_major> qr_util;

```

Comment:
as i am suggesting, pass the dense solver from outside.

@rubenzorrilla can u point to how this is implemented?